src/Shared/FileMatcher.cs

```diff
@@ -1548,14 +1630,30 @@ string fileToMatch
         /// <param name="projectDirectoryUnescaped">The project directory.</param>
         /// <param name="filespecUnescaped">Get files that match the given file spec.</param>
         /// <param name="excludeSpecsUnescaped">Exclude files that match this file spec.</param>
+        /// <param name="entriesCache">Cache used for caching IO operation results</param>
         /// <returns>The array of files.</returns>
         internal static string[] GetFiles
         (
             string projectDirectoryUnescaped,
             string filespecUnescaped,
-            IEnumerable<string> excludeSpecsUnescaped = null
+            IEnumerable<string> excludeSpecsUnescaped = null,
+            ConcurrentDictionary<string, IEnumerable<string>> entriesCache = null
         )
         {
+            var getFileSystemEntries = entriesCache == null
+                ? s_defaultGetFileSystemEntries
+                : (type, path, pattern, directory, projectDirectory) =>
+                {
+                    // Cache only directories, for files we won't ever hit the cache while open a project
+                    if (type == FileSystemEntity.Directories)
+                    {
+                        return entriesCache.GetOrAdd($"{path};{pattern ?? "*"}",
+                            s => new CachedEnumerable<string>(s_defaultGetFileSystemEntries(type, path, pattern,
```

Comment:
Given that `s_defaultGetFileSystemEntries` gets the whole array of entries in one go (https://github.com/Microsoft/msbuild/blob/master/src/Shared/FileMatcher.cs#L256), wouldn't it be better to use something like `ImmutableArray<T>`, initialize with the full array and avoid the locks and contention in `CachedEnumerable`?

Reply:
The main idea of `CachedEnumerable` was to save one iteration through the `IEnumerable<string>` retrived by [Directory.EnumerateDirectory](https://github.com/Microsoft/msbuild/pull/2572/files#diff-64d47c5cbb59959b7b2fa6963672d596R291). After some testing I realized that there are no noticable differences by initialize the `IEnumerable` with `ToImmutableArray` method prior returning. I will remove the `CachedEnumerable` and cache an `ImmutableArray<string>` instead.

Reply:
Oh, but you are switching that to using `EnumerateDirectories` later so this doesn't matter.

Reply:
So, if you are reading all of it here itself then we can switch back to using `Directory.GetDirectories` and to `string[]` from `IEnumerable<string>`.

Also, it would be good to use `git rebase -i` to rework the commits whenever you want to change or fix something in them. That way they look like clean commits and makes it easier to review.

Reply:
Just thinking aloud - `Directory.EnumerateDirectories` should be helping the amount of memory being used at least, in case of *lot* of directories. Didn't it?